File: question5.py
# функция для поиска минимального пути
def find_minimum_path_sum(triangle):
    if not triangle:
        return 0
    
    triangle_height = len(triangle)
    # Используем только один массив длиной triangle_height+1
    dp = [0] * (triangle_height + 1)
    
    # Проходим снизу вверх
    for row in range(triangle_height - 1, -1, -1):
        for col in range(len(triangle[row])):
            # Обновляем на месте - экономим память
            dp[col] = triangle[row][col] + min(dp[col], dp[col + 1])
    
    return dp[0]


# примеры:

triangle1 = [
    [2],
    [3, 4],
    [6, 5, 7],
    [4, 1, 8, 3],
    [6, 4, 9, 1, 7]
]

# Ожидаемый результат: 15
min_sum1 = find_minimum_path_sum(triangle1)
print(f"Треугольник1: Минимальная сумма пути = {min_sum1}") 
# Вывод: 15

triangle2 = [
    [-10]
]

# Ожидаемый результат: -10
min_sum2 = find_minimum_path_sum(triangle2)
print(f"Треугольник2: Минимальная сумма пути = {min_sum2}")
# Вывод: -10

triangle3 = [
    [1],
    [2, 3]
]

# Ожидаемый результат: 3
min_sum3 = find_minimum_path_sum(triangle3)
print(f"Треугольник3: Минимальная сумма пути = {min_sum3}")
# Вывод: 3

File: java/question5.java
package java;
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class question5 {
    public static int findMinimumPathSum(List<List<Integer>> triangle) {
        if (triangle == null || triangle.isEmpty()) {
            return 0;
        }
        
        int triangleHeight = triangle.size();
        // Используем только один массив длиной triangleHeight+1
        int[] dp = new int[triangleHeight + 1];
        
        // Проходим снизу вверх
        for (int row = triangleHeight - 1; row >= 0; row--) {
            for (int col = 0; col < triangle.get(row).size(); col++) {
                // Обновляем на месте - экономим память
                dp[col] = triangle.get(row).get(col) + Math.min(dp[col], dp[col + 1]);
            }
        }
        
        return dp[0];
    }
    
    public static void main(String[] args) {
        // Пример 1
        List<List<Integer>> triangle1 = new ArrayList<>();
        triangle1.add(Arrays.asList(2));
        triangle1.add(Arrays.asList(3, 4));
        triangle1.add(Arrays.asList(6, 5, 7));
        triangle1.add(Arrays.asList(4, 1, 8, 3));
        triangle1.add(Arrays.asList(6, 4, 9, 1, 7));

        int minSum1 = findMinimumPathSum(triangle1);
        System.out.println("Треугольник1: Минимальная сумма пути = " + minSum1);
        // Вывод: 15

        // Пример 2
        List<List<Integer>> triangle2 = new ArrayList<>();
        triangle2.add(Arrays.asList(-10));

        int minSum2 = findMinimumPathSum(triangle2);
        System.out.println("Треугольник2: Минимальная сумма пути = " + minSum2);
        // Вывод: -10

        // Пример 3
        List<List<Integer>> triangle3 = new ArrayList<>();
        triangle3.add(Arrays.asList(1));
        triangle3.add(Arrays.asList(2, 3));

        int minSum3 = findMinimumPathSum(triangle3);
        System.out.println("Треугольник3: Минимальная сумма пути = " + minSum3);
        // Вывод: 3
    }
}
